fix(operator): clean up nodes that no longer exist from status

t0mmylam opened the pull request:
Closes #59 

### Summary
This PR implements functionality to clean up stale node entries from the Skyhook status when nodes no longer exist in the cluster or no longer match the node selector.

### Problem
Currently, the Skyhook status retains entries for nodes that have been removed from the cluster.

### Solution
Added a new `CleanupRemovedNodes` function that:
- Compares nodes present in the cluster against nodes in the Skyhook status
- Removes entries for nodes that no longer exist
- Updates the Skyhook object to reflect changes
- Returns a boolean indicating whether cleanup was performed

The function is called during the reconciliation process to ensure status stays synchronized with the actual cluster state.

### Testing
Tested changes locally by creating a cluster, removing a worker node, and comparing the output of `kubectl describe skyhooks.skyhook.nvidia.com/demo` before and after node removal.

**Unit Tests:**
```
❯ go test -v ./internal/controller -ginkgo.focus="CleanupRemovedNodes"
=== RUN   TestGenerateValidPodNames
--- PASS: TestGenerateValidPodNames (0.00s)
=== RUN   TestControllers
Running Suite: Controller Suite - /Users/tommylam/repos/skyhook/operator/internal/controller
============================================================================================
Random Seed: 1751831957
Will run 2 of 54 specs
SSSSS••SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Ran 2 of 54 Specs in 3.426 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 52 Skipped
--- PASS: TestControllers (3.43s)
PASS
ok  	github.com/NVIDIA/skyhook/internal/controller	3.847s
```
**NOTE:** Occasionally see transient errors in AfterSuite, but tests pass successfully. This appears to be related to test environment cleanup and doesn't affect the functionality, but unsure and just wanted to point it out.

t0mmylam committed:
fix(operator): clean up nodes that no longer exist from status

lockwobr commented on operator/internal/controller/cluster_state_v2_test.go:
```diff
@@ -172,3 +173,119 @@ var _ = Describe("BuildState ordering", func() {
+	})
+})
+
+type mockSkyhookNode struct {
+	wrapper.SkyhookNode // embed the interface
+	nodeName            string
+}
+
+func (m *mockSkyhookNode) GetNode() *corev1.Node {
+	return &corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{Name: m.nodeName},
+	}
+}
```
actually it already exists
https://github.com/NVIDIA/skyhook/blob/main/operator/internal/controller/mock/SkyhookNodes.go

t0mmylam commented on operator/internal/controller/cluster_state_v2.go:
```diff
@@ -668,3 +673,73 @@ func (skyhook *skyhookNodes) Migrate(logger logr.Logger) error {
+	}
+
+	// Check and remove nodes from NodeState that no longer exist
+	if skyhook.GetSkyhook().Status.NodeState != nil {
```
Thanks for the suggestion! I added a helper function and I'll keep this in mind going forward.

t0mmylam commented:
Addressed comments thanks for the feedback! Sorry about the weird commit history, ran into some annoying git stuff between rebasing and the fork, I'll make sure to squash before merging.

lockwobr reviewed (changes requested):
The e2e tests seem to be having issues, you should be able to install kind on your local machine and run the test to see if you track down why they are failing. Once you have [kind](https://kind.sigs.k8s.io/) installed, there are make commands to help use it. Check out `make help`.

t0mmylam commented:
I know you mentioned that ConfigUpdates should be cleaned up in the original issue desc, but it seems like that field is at a package level which is different than all the other ones. That's why the e2e tests were failing so I just removed the config branch and it passed the e2e tests.

t0mmylam commented:
Seems like CI is passing. The push error seems perm based not related to code changes was wondering if there's something I can do on my end to resolve this? @lockwobr
